File: go/lesson-01-introduction/src/main.go
package main

import (
	"fmt"
	"strconv"
)

func main() {
	helloWorld()
	fmt.Println("---")
	showVariables()
	fmt.Println("---")
	showControlFlow()
	fmt.Println("---")
	showDefer()
}

func helloWorld() {
	fmt.Println("Hello, World!")
	fmt.Printf("Hello, %s! You are %d years old.\n", "Alice", 30)
	fmt.Printf("Pi is approximately %.4f\n", 3.14159)
}

func showVariables() {
	// var declaration — explicit type
	var age int = 25
	var name string = "Alice"
	var price float64 = 9.99
	var isActive bool = true

	// Short declaration — type is inferred
	city := "Hanoi"
	score := 95.5

	// Constant
	const MaxRetries = 3
	const Greeting = "Hello"

	// Zero values — declared without assignment
	var zeroInt int        // 0
	var zeroStr string     // ""
	var zeroBool bool      // false

	fmt.Printf("age=%d name=%s price=%.2f active=%v\n", age, name, price, isActive)
	fmt.Printf("city=%s score=%.1f\n", city, score)
	fmt.Printf("MaxRetries=%d Greeting=%s\n", MaxRetries, Greeting)
	fmt.Printf("zero values: int=%d string=%q bool=%v\n", zeroInt, zeroStr, zeroBool)

	// Multiple assignment
	x, y := 10, 20
	fmt.Printf("before swap: x=%d y=%d\n", x, y)
	x, y = y, x
	fmt.Printf("after swap:  x=%d y=%d\n", x, y)

	// Explicit type conversion
	n := 42
	f := float64(n)
	s := strconv.Itoa(n) // int → string
	fmt.Printf("int=%d float=%f string=%q\n", n, f, s)
}

func showControlFlow() {
	score := 85

	// if / else if / else
	if score >= 90 {
		fmt.Println("Grade: A")
	} else if score >= 80 {
		fmt.Println("Grade: B")
	} else if score >= 70 {
		fmt.Println("Grade: C")
	} else {
		fmt.Println("Grade: F")
	}

	// if with init statement — n is scoped to the if block
	if n, err := strconv.Atoi("42"); err == nil {
		fmt.Printf("Parsed number: %d\n", n)
	}

	// for — Go's only loop
	fmt.Print("for loop: ")
	for i := 0; i < 5; i++ {
		fmt.Printf("%d ", i)
	}
	fmt.Println()

	// while-style for
	fmt.Print("while-style: ")
	count := 0
	for count < 3 {
		fmt.Printf("%d ", count)
		count++
	}
	fmt.Println()

	// range over a string (gives index and rune)
	fmt.Print("range string: ")
	for i, ch := range "Go" {
		fmt.Printf("[%d]%c ", i, ch)
	}
	fmt.Println()

	// switch — no fallthrough by default
	day := "Mon"
	switch day {
	case "Mon", "Tue", "Wed", "Thu", "Fri":
		fmt.Println(day, "is a weekday")
	case "Sat", "Sun":
		fmt.Println(day, "is the weekend")
	default:
		fmt.Println("unknown day")
	}

	// switch without expression — acts like if/else chain
	switch {
	case score >= 90:
		fmt.Println("Excellent")
	case score >= 80:
		fmt.Println("Good")
	default:
		fmt.Println("Keep trying")
	}
}

func showDefer() {
	fmt.Println("defer example:")
	defer fmt.Println("  3: deferred — runs last (LIFO)")
	defer fmt.Println("  2: deferred — runs second-to-last")
	fmt.Println("  1: runs first")
	// Output order: 1, 2, 3
}

File: nodejs/lesson-01-introduction/src/main.js
'use strict';

showVariables();
console.log('---');
showTypes();
console.log('---');
showControlFlow();

// ─── Variables ────────────────────────────────────────────────────────────────

function showVariables() {
  console.log('=== Variables ===');

  // const — cannot be reassigned (use this by default)
  const name = 'Alice';
  const age = 30;
  const PI = 3.14159;

  // let — block-scoped, can be reassigned
  let score = 85;
  score = 90;

  // Template literals — preferred over string concatenation
  console.log(`Hello, ${name}! You are ${age} years old.`);
  console.log(`Score: ${score}, PI: ${PI.toFixed(2)}`);

  // Destructuring assignment
  const [first, second, ...rest] = [10, 20, 30, 40, 50];
  console.log(`first=${first} second=${second} rest=${rest}`);

  const { city = 'unknown', country } = { country: 'Vietnam' };
  console.log(`city=${city} country=${country}`);

  // Multiple assignment
  let x = 1, y = 2;
  [x, y] = [y, x]; // swap
  console.log(`after swap: x=${x} y=${y}`);
}

// ─── Types ────────────────────────────────────────────────────────────────────

function showTypes() {
  console.log('=== Data Types & Coercion ===');

  // Primitives
  const str = 'hello';
  const num = 42;
  const float = 3.14;
  const bool = true;
  const nothing = null;
  let undef; // undefined
  const big = 9007199254740991n; // BigInt

  console.log(typeof str, typeof num, typeof bool);
  console.log('typeof null:', typeof null); // 'object' ← JS quirk
  console.log('typeof undef:', typeof undef);
  console.log('Array.isArray([]):', Array.isArray([]));

  // Falsy values — all coerce to false
  const falsy = [0, '', null, undefined, NaN, false];
  falsy.forEach(v => console.log(`  ${String(v)} is falsy:`, !v));

  // === strict vs == loose equality
  console.log("1 == '1':", 1 == '1');    // true (coerced)
  console.log("1 === '1':", 1 === '1'); // false (strict)
  console.log('null == undefined:', null == undefined);  // true
  console.log('null === undefined:', null === undefined); // false

  // Type conversion
  console.log('Number("42"):', Number('42'));
  console.log('Number(""):', Number(''));   // 0
  console.log('Number("abc"):', Number('abc')); // NaN
  console.log('String(42):', String(42));
  console.log('Boolean(0):', Boolean(0)); // false
  console.log('Boolean("x"):', Boolean('x')); // true
  console.log('parseInt("42px"):', parseInt('42px')); // 42

  // Optional chaining & nullish coalescing
  const user = { profile: { city: 'Hanoi' } };
  console.log('user?.profile?.city:', user?.profile?.city);     // Hanoi
  console.log('user?.address?.city:', user?.address?.city);     // undefined
  console.log('port ?? 3000:', undefined ?? 3000);               // 3000
  console.log('port ?? 3000 (0):', 0 ?? 3000);                  // 0 (not null/undef)
}

// ─── Control Flow ─────────────────────────────────────────────────────────────

function showControlFlow() {
  console.log('=== Control Flow ===');

  const score = 85;

  // if / else if / else
  if (score >= 90) {
    console.log('Grade: A');
  } else if (score >= 80) {
    console.log('Grade: B');
  } else {
    console.log('Grade: C');
  }

  // Ternary
  const label = score >= 90 ? 'Pass (excellent)' : score >= 60 ? 'Pass' : 'Fail';
  console.log('Label:', label);

  // for loop
  process.stdout.write('for loop: ');
  for (let i = 0; i < 5; i++) process.stdout.write(i + ' ');
  console.log();

  // for...of — iterate values
  process.stdout.write('for...of: ');
  for (const fruit of ['apple', 'banana', 'cherry']) {
    process.stdout.write(fruit + ' ');
  }
  console.log();

  // while
  let count = 0;
  while (count < 3) {
    process.stdout.write(count + ' ');
    count++;
  }
  console.log('(while)');

  // switch
  const day = 'Mon';
  switch (day) {
    case 'Mon':
    case 'Tue':
    case 'Wed':
    case 'Thu':
    case 'Fri':
      console.log(day, 'is a weekday');
      break;
    case 'Sat':
    case 'Sun':
      console.log(day, 'is a weekend');
      break;
    default:
      console.log('unknown day');
  }

  // Short-circuit evaluation
  const user = null;
  console.log('user?.name:', user?.name); // undefined (no error)

  const config = {};
  config.timeout ??= 5000; // assign only if null/undefined
  console.log('timeout:', config.timeout); // 5000
}
